OpenRA.Mods.Common/Widgets/ObserverArmyIconsWidget.cs

```diff
@@ -0,0 +1,209 @@
+
+			MinWidth = other.MinWidth;
+
+			TooltipUnit = other.TooltipUnit;
+			GetTooltipUnit = () => TooltipUnit;
+
+			TooltipTemplate = other.TooltipTemplate;
+			TooltipContainer = other.TooltipContainer;
+
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+		}
+
+		public override void Draw()
+		{
+			armyIcons.Clear();
+
+			var player = GetPlayer();
+			if (player == null)
+				return;
+
+			var playerStatistics = stats.Update(player);
+
+			var items = playerStatistics.Units.Values
+				.Where(u => u.Count > 0 && u.Icon != null)
+				.OrderBy(u => u.ProductionQueueOrder)
+				.ThenBy(u => u.BuildPaletteOrder);
+
+			Game.Renderer.EnableAntialiasingFilter();
+
+			var queueCol = 0;
+			foreach (var unit in items)
+			{
+				var icon = unit.Icon;
+				var topLeftOffset = new int2(queueCol * (IconWidth + IconSpacing), 0);
+
+				var iconTopLeft = RenderOrigin + topLeftOffset;
+				var centerPosition = iconTopLeft;
+
+				WidgetUtils.DrawSHPCentered(icon.Image, centerPosition + 0.5f * iconSize, worldRenderer.Palette(unit.IconPalette), 0.5f);
+
+				armyIcons.Add(new ArmyIcon
+				{
+					Bounds = new Rectangle(iconTopLeft.X, iconTopLeft.Y, (int)iconSize.X, (int)iconSize.Y),
+					Unit = unit
+				});
+
+				queueCol++;
+			}
+
+			Bounds.Width = queueCol * (IconWidth + IconSpacing);
+
+			Game.Renderer.DisableAntialiasingFilter();
+
+			var bold = Game.Renderer.Fonts["TinyBold"];
+			foreach (var armyIcon in armyIcons)
+			{
+				var text = armyIcon.Unit.Count.ToString();
+				bold.DrawTextWithContrast(text, armyIcon.Bounds.Location + new float2(iconSize.X, 0) - new float2(bold.Measure(text).X, bold.TopOffset),
+					Color.White, Color.Black, 1);
+			}
+
+			var parentWidth = Bounds.X + Bounds.Width;
+			Parent.Bounds.Width = parentWidth;
+
+			var gradient = Parent.Get<GradientColorBlockWidget>("PLAYER_GRADIENT");
+
+			var offset = gradient.Bounds.X - Bounds.X;
+			var gradientWidth = Math.Max(MinWidth - offset, queueCol * (IconWidth + IconSpacing));
+
+			gradient.Bounds.Width = gradientWidth;
+			var widestChildWidth = Parent.Parent.Children.Max(x => x.Bounds.Width);
+
+			Parent.Parent.Bounds.Width = Math.Max(25 + widestChildWidth, Bounds.Left + MinWidth);
+		}
+
+		public override Widget Clone()
+		{
+			return new ObserverArmyIconsWidget(this);
+		}
+
+		public override void MouseEntered()
+		{
+			if (TooltipContainer == null)
+				return;
+
+			foreach (var armyIcon in armyIcons)
+			{
+				if (!armyIcon.Bounds.Contains(Viewport.LastMousePos))
+					continue;
+
+				TooltipUnit = armyIcon.Unit;
+				break;
+			}
+
+			tooltipContainer.Value.SetTooltip(TooltipTemplate, new WidgetArgs { { "getTooltipUnit", GetTooltipUnit } });
+		}
+
+		public override void MouseExited()
+		{
+			if (TooltipContainer == null)
+				return;
+
+			tooltipContainer.Value.RemoveTooltip();
+		}
+
+		public override void Tick()
+		{
+			if (lastIconIdx >= armyIcons.Count)
```

Comment:
I don't know what this is for. This prevents me from getting any more tooltips if I hovered over an icon with an index bigger or equal to the current amount of icons.